### PR TITLE
feat: implement first version of osmosis v16 pool

### PR DESCRIPTION
**Pull request by bkioshn**
## Describe your changes
Implement first version of Osmosis v16 pool (Concentrated liquidity pool and CosmWasm pool)

### CODE REVIEW

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 1]
```diff
@@ -57,47 +57,75 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
...
+            {formatRatio(Number(pool.exitFee) as Ratio<number>)}
+          </Text>
+        </LabelText>
+      )}
+      {pool.futurePoolGovernor !== null && (
+        <LabelText label="Future Governor">
+          {futurePoolGovernorType !== "invalid_address" ? (
+            <ExplorerLink
+              type={futurePoolGovernorType}
+              value={pool.futurePoolGovernor ?? "N/A"}
```
Comment (evilpeach) [inline src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:86]: remove `?? "N/A"`

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 2]
```diff
@@ -57,47 +57,75 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
...
+              isReadOnly={!pool.futurePoolGovernor}
```
Comment (evilpeach) [inline src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:87]: How about this one? remove?

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 3]
```diff
@@ -57,47 +57,75 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
...
+              textFormat="truncate"
+              w="140px"
+            />
+          ) : (
+            <Text
+              as="p"
+              color={pool.futurePoolGovernor ? "text.main" : "text.disabled"}
+              variant="body2"
+            >
+              {pool.futurePoolGovernor ?? "N/A"}
```
Comment (evilpeach) [inline src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:98]: remove `??`

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 4]
```diff
@@ -57,47 +57,75 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
...
+              value={pool.futurePoolGovernor ?? "N/A"}
+              isReadOnly={!pool.futurePoolGovernor}
+              showCopyOnHover
+              textFormat="truncate"
+              w="140px"
+            />
+          ) : (
+            <Text
+              as="p"
+              color={pool.futurePoolGovernor ? "text.main" : "text.disabled"}
```
Comment (evilpeach) [inline src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:95]: I think it's no need to handle `pool.futurePoolGovernor` is undefined.

File: src/lib/pages/pools/poolId.tsx [hunk 5]
```diff
@@ -23,7 +24,13 @@ export const PoolId = () => {
...
   useEffect(() => {
     if (router.isReady) AmpTrack(AmpEvent.TO_POOL_DETAIL);
-  }, [router.isReady]);
+    // First version, navigate to contract details page if pool type is CosmWasm
+    if (pool?.type === PoolType.COSMWASM)
+      navigate({
+        pathname: `/contracts/${pool.contractAddress}`,
+        replace: true,
+      });
+  }, [navigate, pool, pool?.contractAddress, pool?.type, router.isReady]);
```
Comment (evilpeach) [inline src/lib/pages/pools/poolId.tsx:33]: remove `pool`

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/pages/pools/components/unsupportedSection/UnsupportedPoolCard.tsx [hunk 6]
```diff
@@ -113,67 +114,77 @@ export const UnsupportedPoolCard = ({ item }: UnsupportedPoolCardProps) => {
...
-              </Text>
-              <Flex gap={4} flexDirection="column">
-                <Flex gap={2} flexDirection="column">
-                  {item.poolLiquidity.map((asset) => (
-                    <Flex
-                      className="copier-wrapper"
-                      key={asset.denom}
-                      gap={3}
-                      alignItems="center"
+            {item.poolLiquidity.some((coin) => !coin.amount) ? (
```
Comment (evilpeach) [inline src/lib/pages/pools/components/unsupportedSection/UnsupportedPoolCard.tsx:117]: add `TODO: revisit later`

File: src/lib/pages/pools/components/FilterByPoolType.tsx [hunk 7]
```diff
@@ -21,21 +22,33 @@ interface FilterByPoolTypeProps {
...
     label: "StableSwap Pools",
-    value: "Stableswap",
+    value: PoolType.STABLESWAP,
     image: STABLESWAP_ICON,
     disabled: false,
   },
+  {
+    label: "Concentrated Liquidity Pools",
+    value: PoolType.CL,
+    image: BALANCER_ICON,
```
Comment (evilpeach) [inline src/lib/pages/pools/components/FilterByPoolType.tsx:43]: Is it correct to use this icon?

File: src/lib/pages/pools/components/PoolHeader.tsx [hunk 8]
```diff
@@ -12,100 +18,123 @@ interface PoolHeaderProps
...
-              <Flex alignItems="center" gap={1}>
-                <Image boxSize={4} src={SUPERFLUID_ICON} />
-                <Text variant="body2" color="text.dark">
-                  Superfluid
-                </Text>
+            )}
+            {isSuperfluid && (
+              <Flex alignItems="center" gap={2}>
+                <Flex
+                  backgroundColor="pebble.600"
```
Comment (evilpeach) [inline src/lib/pages/pools/components/PoolHeader.tsx:122]: primary.600

File: src/lib/pages/pools/components/supportedSection/PoolCard.tsx [hunk 9]
```diff
@@ -89,60 +92,69 @@ export const PoolCard = ({ item, mode = "percent-value" }: PoolCardProps) => {
...
-                    big(0)
-                  ) as USD<Big>
-              }
-              liquidity={liquidity}
-              mode={mode}
-            />
-          )}
-        </>
-      </SimpleGrid>
+      {item.poolLiquidity.some((coin) => !coin.amount) ? (
```
Comment (evilpeach) [inline src/lib/pages/pools/components/supportedSection/PoolCard.tsx:108]: add `TODO: revisit later`

File: src/lib/utils/assetValue.ts [hunk 10]
```diff
@@ -41,20 +41,21 @@ export const coinToTokenWithValue = (
   amount: string,
   assetInfo: Option<AssetInfo>
 ): TokenWithValue => {
-  const tokenAmount = big(amount) as U<Token<Big>>;
+  const tokenAmount = (amount.length === 0 ? 0 : big(amount)) as U<Token<Big>>;
```
Comment (evilpeach) [inline src/lib/utils/assetValue.ts:44]: add `TODO: ...`

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 11]
```diff
@@ -57,47 +57,75 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
...
-            w="140px"
-          />
-        ) : (
-          <Text
-            as="p"
-            color={pool.futurePoolGovernor ? "text.main" : "text.disabled"}
-            variant="body2"
-          >
-            {pool.futurePoolGovernor ?? "N/A"}
+      <Divider orientation="vertical" h="46px" />
```
[the remark this answers is not in the document]
Comment (bkioshn) [reply src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:60]: somehow using eg {12} doesn't work

File: src/lib/pages/pools/components/FilterByPoolType.tsx (hunk 7 above)
Comment (bkioshn) [reply src/lib/pages/pools/components/FilterByPoolType.tsx:43]: yes, the same as Balancer pool

Comment (evilpeach) [reply src/lib/pages/pools/components/FilterByPoolType.tsx:43]: okay

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (evilpeach) [reply src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:60]: Ah, it's okay

File: src/lib/pages/pools/components/supportedSection/PoolCard.tsx [hunk 12]
```diff
@@ -31,7 +34,14 @@ export const PoolCard = ({ item, mode = "percent-value" }: PoolCardProps) => {
 
   const navigate = useInternalNavigate();
   const handleOnClick = () => {
-    navigate({ pathname: `/pools/[poolId]`, query: { poolId: item.id } });
+    // First version, navigate to contract details page if pool type is CosmWasm
+    if (item?.type === PoolType.COSMWASM && item.contractAddress)
+      navigate({
```
Comment (evilpeach) [inline src/lib/pages/pools/components/supportedSection/PoolCard.tsx:39]:
```tsx
      navigate({
        pathname: `/contracts/[contractAddress]`,
        query: { contractAddress: item.contractAddress },
      });
```

File: src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx [hunk 13]
```diff
@@ -78,24 +78,19 @@ export const PoolInfo = ({ pool }: PoolInfoProps) => {
           </Text>
         </LabelText>
       )}
-      {pool.futurePoolGovernor !== null && (
+      {pool.futurePoolGovernor && (
```
Comment (evilpeach) [inline src/lib/pages/pools/components/pool-details/header/PoolInfo.tsx:81]: Let's use `!== null` as before.

File: src/lib/pages/pools/components/unsupportedSection/UnsupportedPoolCard.tsx [hunk 14]
```diff
@@ -47,7 +48,15 @@ export const UnsupportedPoolCard = ({ item }: UnsupportedPoolCardProps) => {
 
   const navigate = useInternalNavigate();
   const handleOnClick = () => {
-    navigate({ pathname: `/pools/[poolId]`, query: { poolId: item.id } });
+    // First version, navigate to contract details page if pool type is CosmWasm
+    if (item?.type === PoolType.COSMWASM && item.contractAddress)
+      navigate({
+        pathname: `/contracts/${item.contractAddress}`,
+        replace: true,
```
Comment (evilpeach) [inline src/lib/pages/pools/components/unsupportedSection/UnsupportedPoolCard.tsx:55]: change this as well

File: src/lib/utils/assetValue.ts [hunk 15]
```diff
@@ -41,6 +41,9 @@ export const coinToTokenWithValue = (
   amount: string,
   assetInfo: Option<AssetInfo>
 ): TokenWithValue => {
+  /**
+   * @remarks Assuming if liquidty amount does not exist, it will result in empty string
+   */
   const tokenAmount = (amount.length === 0 ? 0 : big(amount)) as U<Token<Big>>;
```
Comment (evilpeach) [inline src/lib/utils/assetValue.ts:47]:
```tsx
const tokenAmount = (amount.length === 0 ? big(0) : big(amount)) as U<Token<Big>>;
```

**Review by evilpeach — APPROVED** [review]

**Pull request merged**